### PR TITLE
Add FalkordbRM Retriever Class for Enhanced Querying and Embedding Retrieval

### PR DESCRIPTION
**Pull request by kingtroga**
This pull request introduces the FalkordbRM retriever class, designed to integrate FalkorDB for efficient querying and passage retrieval based on embeddings. The class supports both local and cloud-based FalkorDB sessions and uses OpenAI's embedding services by default

**Key Changes**:
1. FalkordbRM Retriever Class
2. Embedding Provider Integration
3. Error Handling
4. Customization Options

### CODE REVIEW

File: dspy/retrieve/falkordb_rm.py [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+
+        for vector in query_vectors:
+            params = {
+                "embedding": vector,
+                "node_label": self.node_label,
+                "text_node_property": self.text_node_property,
+                "embedding_node_property": self.embedding_node_property,
+                "k": k,
+            }
+            print(params)
```
Comment (galshubeli) [inline dspy/retrieve/falkordb_rm.py:191]: do we need this print?

File: docs/docs/deep-dive/retrieval_models_clients/FalkordbRM.md [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+
+ex:
+
+```python
+Prediction(
+    passages=['Passage 1 Lorem Ipsum awesom', 'Passage 2 Lorem Ipsum Youppidoo', 'Passage 3 Lorem Ipsum Yassssss']
+)
+```
+
+### Quick Example how to use Neo4j in a local environment.
```
Comment (galshubeli) [inline docs/docs/deep-dive/retrieval_models_clients/FalkordbRM.md:69]: falkorDB?

File: docs/docs/deep-dive/retrieval_models_clients/FalkordbRM.md [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+- `retrieval_query` (_str_, _optional_): A custom query string provided for retrieving data. If not provided, a default query tailored to the `text_node_property` will be used.
+- `embedding_provider` (_str_, _optional_): The name of the service provider for generating embeddings. Only "openai" is supported.
+- `embedding_model` (_str_, _optional_): The specific embedding model to use from the provider. By default, it uses the "text-embedding-ada-002" model from OpenAI.
+
+
+### Methods
+
+#### `forward(self, query: [str], k: Optional[int] = None) -> dspy.Prediction`
+
+Search the neo4j vector index for the top `k` passages matching the given query or queries, using embeddings generated via the specified `embedding_model`.
```
Comment (galshubeli) [inline docs/docs/deep-dive/retrieval_models_clients/FalkordbRM.md:50]: falkordb

File: docs/docs/deep-dive/retrieval_models_clients/FalkordbRM.md (hunk 2 above)
Comment (kingtroga) [reply docs/docs/deep-dive/retrieval_models_clients/FalkordbRM.md:69]: Changed

File: docs/docs/deep-dive/retrieval_models_clients/FalkordbRM.md (hunk 3 above)
Comment (kingtroga) [reply docs/docs/deep-dive/retrieval_models_clients/FalkordbRM.md:50]: Changed!

**Comment by gkorland** [comment]
@okhat @arnavsinghvi11 can you please merge?

File: dspy/retrieve/falkordb_rm.py [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+
+from dspy import Retrieve, Prediction
+from dsp.utils.settings import settings
+from dsp.utils import dotdict
+
+try:
+    import falkordb
+except ImportError:
+    raise ImportError(
+        "Please install the falkordb package by running `pip install dspy-ai[falkordb]`"
```
Comment (arnavsinghvi11) [inline dspy/retrieve/falkordb_rm.py:23]: this actually will require adding the `falkordb` related dependencies to the extras [here](https://github.com/stanfordnlp/dspy/blob/89925ff6ab049726b310a3f80dda3581fe6d6ef4/setup.py)

Feel free to reference the [other supported extras ](https://dspy-docs.vercel.app/quick-start/installation/#__tabbed_1_2) as well for related context

**Comment by arnavsinghvi11** [comment]
Thanks @kingtroga @gkorland  for this PR! Seems like some merge conflicts should be resolved if you pull from current main! should pass the failing test as well

also added a comment to support `pip install dspy-ai[falkordb]` correctly.

**Review by galshubeli — APPROVED** [review]

File: dspy/retrieve/falkordb_rm.py (hunk 4 above)
Comment (kingtroga) [reply dspy/retrieve/falkordb_rm.py:23]: done. thank you

**Comment by kingtroga** [comment]
@arnavsinghvi11 I've made the necessary changes. Please try the merge again 🙏🏾

**Comment by kingtroga** [comment]
@arnavsinghvi11

**Comment by arnavsinghvi11** [comment]
Thanks @kingtroga !

we're currently planning an internal refactor of how we support RMs in DSPy and hence will leave this PR open to avoid any future code reversions. However, if the PR requires merging to be usable in DSPy, we'll do so! Please let us know

**Comment by kingtroga** [comment]
Hi @arnavsinghvi11 , thanks for the update! This PR does need merging to be usable in DSPy. Let me know if there's anything I can do to help with it.

**Comment by arnavsinghvi11** [comment]
Thanks @kingtroga !

**Pull request merged**